{
    "id": "LUCENE-1397",
    "title": "When BG merge hits an exception, optimize sometimes throws an IOException missing the root cause",
    "details": {
        "labels": "",
        "priority": "Minor",
        "components": [
            "core/index"
        ],
        "type": "Bug",
        "fix_versions": [
            "2.4",
            "2.9"
        ],
        "affect_versions": "2.4",
        "resolution": "Fixed",
        "status": "Closed"
    },
    "description": "When IndexWriter.optimize() is called, ConcurrentMergeScheduler will\nrun the requested merges with background threads and optimize() will\nwait for these merges to complete.\n\nIf a merge hits an exception, it records the root cause exception such\nthat optimize can then retrieve this root cause and throw its own\nexception, with the root cause.\n\nBut there is a bug: sometimes, the fact that an exception occurred on\na merge is recorded, but the root cause is missing.  In this cause,\noptimize() still throws an exception (correctly indicating that the\noptimize() has not finished successfully), but it's not helpful\nbecause it's missing the root cause.  You must then go find the root\ncause in the JRE's stderr logs.\n\nThis has hit a few users on this lists, most recently:\n\n  http://www.nabble.com/Background-merge-hit-exception-td19540409.html#a19540409\n\nI found the isssue, and finally got a unit test to intermittently show\nit.  It's a simple thread safety issue: in a finally clause in\nIndexWriter.merge we record the fact that the merge hit an exception\nbefore actually setting the root cause, and then only in\nConcurrentMergeScheduler's exception handler do we set the root\ncause.  If the optimize thread is scheduled in between these two, it\ncan throw an exception missing its root cause.\n\nThe fix is straightforward.  I plan to commit to 2.4 & 2.9.",
    "attachments": {
        "LUCENE-1397.patch": "https://issues.apache.org/jira/secure/attachment/12390569/LUCENE-1397.patch"
    },
    "issue_links": {},
    "comments": [
        {
            "date": "2008-09-20T10:02:33+0000",
            "content": "Attached patch.  I plan to commit in a day or so, and then roll another RC. ",
            "author": "Michael McCandless",
            "id": "comment-12632930"
        },
        {
            "date": "2008-09-22T22:56:12+0000",
            "content": "Committed revision 698025 (trunk) and 698026 (2.4) ",
            "author": "Michael McCandless",
            "id": "comment-12633547"
        }
    ]
}